apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/KotlinCodeGen.kt

```diff
@@ -331,25 +347,31 @@ internal object KotlinCodeGen {
     }
   }
 
-  private val FieldType.writeListItemCode: CodeBlock
+  private val FieldType.Array.writeListItemCode: CodeBlock
     get() {
-      return when (this) {
-        is FieldType.Scalar -> when (this) {
+      val safeValue = if (isOptional) "value?" else "value"
+      return when (rawType) {
+        is FieldType.Scalar -> when (rawType) {
           is FieldType.Scalar.String -> CodeBlock.of("listItemWriter.writeString(value)")
           is FieldType.Scalar.Int -> CodeBlock.of("listItemWriter.writeInt(value)")
           is FieldType.Scalar.Boolean -> CodeBlock.of("listItemWriter.writeBoolean(value)")
           is FieldType.Scalar.Float -> CodeBlock.of("listItemWriter.writeDouble(value)")
-          is FieldType.Scalar.Enum -> CodeBlock.of("listItemWriter.writeString(value?.rawValue)")
-          is FieldType.Scalar.Custom -> CodeBlock.of("listItemWriter.writeCustom(%T.%L, value)", customEnumType.asTypeName(),
-              customEnumConst)
+          is FieldType.Scalar.Enum -> CodeBlock.of("listItemWriter.writeString($safeValue.rawValue)")
+          is FieldType.Scalar.Custom -> {
+            CodeBlock.of("listItemWriter.writeCustom(%T.%L, value)", rawType.customEnumType.asTypeName(), rawType.customEnumConst)
+          }
         }
-        is FieldType.Object -> CodeBlock.of("listItemWriter.writeObject(value?.marshaller())", asTypeName())
+        is FieldType.Object -> CodeBlock.of("listItemWriter.writeObject($safeValue.marshaller())", asTypeName())
         is FieldType.Array -> {
           CodeBlock.builder()
               .beginControlFlow(
                   "listItemWriter.writeList(value)·{ value, listItemWriter ->",
-                  List::class.asClassName().parameterizedBy(rawType.asTypeName()))
-              .beginControlFlow("value?.forEach·{ value ->", List::class.asClassName().parameterizedBy(rawType.asTypeName()))
+                  List::class.asClassName().parameterizedBy(rawType.rawType.asTypeName())
+              )
+              .beginControlFlow(
+                  "value?.forEach·{ value ->", // value always nullable in ListItemWriter
```

Comment:
hm, that strange why do we have this in `ResponseWriter`:
```
interface ListWriter<T> {
    fun write(items: List<T>?, listItemWriter: ListItemWriter)
  }
```
I have to double check, as it looks like it should be:
```
interface ListWriter<T> {
    fun write(items: List<T>, listItemWriter: ListItemWriter)
  }
```

Reply:
Would be good to merge this PR? I was wondering about this too. It would be great if you can take a look after merging this.

Reply:
Yeah, pls merge will look after